Add authentication unit tests

dknupp opened the pull request:
Create a set of simple unit tests to confirm authentication works.
These can serve as a way to guarantee that any future development
or refactoring won't introduce regressions.

No other code was changed in ths PR. This is only new test code.

Note that two of the tests are strict unit tests and rely on mocking
to confirm proper behavior. The remaining tests require an actual
live deployment, and some special tooling. If these requirements
can't be met, then the tests will either be skipped or will xfail.

Sample test run:
```
% DOMINO_HOST_KEY_NAME=<redacted> DOMINO_API_KEY_KEY_NAME=<redacted> pytest -sv tests/test_basic_auth.py
====================================== test session starts ======================================
cachedir: .pytest_cache
rootdir: /path/to/python-domino, configfile: pytest.ini
plugins: requests-mock-1.9.3, anyio-3.1.0
collected 4 items

tests/test_basic_auth.py::test_object_creation_with_api_key PASSED
tests/test_basic_auth.py::test_object_creation_with_token_file PASSED
tests/test_basic_auth.py::test_auth_against_real_deployment_with_api_key PASSED
tests/test_basic_auth.py::test_auth_against_real_deployment_with_token_file PASSED
```

ddl-alexpanin commented on tests/test_basic_auth.py:
```diff
@@ -0,0 +1,143 @@
+@pytest.fixture
+def real_token_file(tmpdir):
+    """
+    Derive a real authentication token (assumes that certain packages are available).
```
Does it make sense to bring these packages in? Trying to understand long term strategy here.

dknupp replied:
It's an internal package that's a) still under development, and b) probably won't be made available for public consumption anyway. The thinking is that I wanted these tests to be available for internal use, but not cause outright failures (i.e., false negatives) in case all of the test requirements aren't available if by an external contributor.

Here's an example of what the results look like without sample skipped and xfail'ed tests.
```
tests/test_basic_auth.py::test_object_creation_with_api_key PASSED
tests/test_basic_auth.py::test_object_creation_with_token_file PASSED
tests/test_basic_auth.py::test_auth_against_real_deployment_with_api_key SKIPPED (No API key in environment)
tests/test_basic_auth.py::test_auth_against_real_deployment_with_token_file XFAIL (Could not derive valid token for token file)
```

mikesbrown commented on tests/test_basic_auth.py:
```diff
@@ -0,0 +1,143 @@
+        "version": "9.9.9"
+    }
+
+    with requests_mock.Mocker() as mock_endpoint:
```
I didn't know about `requests_mock`. Very clean to use.

mikesbrown commented on tests/test_basic_auth.py:
```diff
@@ -0,0 +1,143 @@
+@pytest.fixture
+def mock_domino_version_response():
+    """
+    Simulate a valid repsonse to an authenticated /version GET request.
```
```suggestion
    Simulate a valid response to an authenticated /version GET request.
```

mikesbrown commented on tests/test_basic_auth.py:
```diff
@@ -0,0 +1,143 @@
+    """
+    Simulate a token file.
+    """
+    token_file = tmpdir.join("dummy_token_file.txt")
```
I didn't realize `tmpdir` was / had become something this sophisticated: you're not opening a file or anything. I suggeset a refer to it for interested parties.
```suggestion
    # Refer to https://docs.pytest.org/en/6.2.x/tmpdir.html#the-tmpdir-fixture for tmpdir usage
    token_file = tmpdir.join("dummy_token_file.txt")
```

mikesbrown commented on tests/test_basic_auth.py:
```diff
@@ -0,0 +1,143 @@
+    """
+    Confirm that the expected auth type is used when using an api key.
+    """
+    dummy_host = "http://mock.domino.com"
```
Changed to use our actual domain. I realize it doesn't matter for real, but still...
```suggestion
    dummy_host = "http://mock.dominodatalab.com"
```

mikesbrown commented on tests/test_basic_auth.py:
```diff
@@ -0,0 +1,143 @@
+    }
+
+    with requests_mock.Mocker() as mock_endpoint:
+        mock_endpoint.get("http://mock.domino.com/version", json=version_info, status_code=200)
```
Changed to use our real domain.
```suggestion
        mock_endpoint.get("http://mock.dominodatalab.com/version", json=version_info, status_code=200)
```

mikesbrown replied:
But it's not a dummy domain: `domino.com` is someone else's domain. So is `somecompany.com`. Both of these are registered domains. I know you're not making real calls, but since this is OSS, nor do I want bs crawlers to see that and associate that domain with our code. Consider it just wanting not to have weird associations on the Internet.

mikesbrown replied:
Oh, and I missed the `repo.com` usage. I don't know why the `domino.com` stuck out: maybe another reason is it's metasimilar but not our correct domain name?

mikesbrown commented on tests/test_basic_auth.py:
```diff
@@ -0,0 +1,143 @@
+    """
+    Confirm that the expected auth type is used when using a token file.
+    """
+    dummy_host = "http://mock.domino.com"
```
```suggestion
    dummy_host = "http://mock.dominodatalab.com"
```

mikesbrown commented on tests/test_basic_auth.py:
```diff
@@ -0,0 +1,143 @@
+@pytest.mark.skipif(not domino_is_reachable(), reason="No access to a live Domino deployment")
+@pytest.mark.skipif(not os.getenv('DOMINO_API_KEY_KEY_NAME'), reason="No API key in environment")
+def test_auth_against_real_deployment_with_api_key():
+    '''
```
Nit: use consistent docstrings. I also believe some PEP recommends triple-double quotes for docstrings always.
```suggestion
    """
```

mikesbrown commented on tests/test_basic_auth.py:
```diff
@@ -0,0 +1,143 @@
+
+    Assumes that ${DOMINO_HOST_KEY_NAME} contains a valid Domino URL
+    Assumes that ${DOMINO_API_KEY_KEY_NAME} contains a valid API key
+    '''
```
```suggestion
    """
```

mikesbrown commented on tests/test_basic_auth.py:
```diff
@@ -0,0 +1,143 @@
+    '''
+    Confirm against a live system that validating by API key works.
+
+    Assumes that ${DOMINO_HOST_KEY_NAME} contains a valid Domino URL
+    Assumes that ${DOMINO_API_KEY_KEY_NAME} contains a valid API key
```
Here and elsewhere, can the environment variable names align with the rest of the repo?
```suggestion
    Assumes that ${DOMINO_API_HOST} contains a valid Domino URL
    Assumes that ${DOMINO_USER_API_KEY} contains a valid API key
```

dknupp commented:
Thanks for the thorough review @mikesbrown.

I think I've addressed all of your comments. There were a lot of changes, but they were made in several distinct commits, so hopefully they're easy to review in isolation. The summary is this:

- updated the endpoint name for the `/version` mock
- corrected usage of constants used as environment variable names
- refactored URL parsing to derive an FQDN
- realized the need for a new fixture to clean the environment for API key tests
- per our discussion, removed the fixture for generating a token file that relied on internal code
- it's incumbent upon testers to provide a valid API key and or token file for tests that require them
- if a valid API key and or token file is not provided, tests are skipped, rather than failing

The last edit was that I realized I need to add an actual API call to validate proper authentication. I chose `environments_list` because it actually doesn't even require a valid user or valid project name to succeed.

dknupp commented:
Here's an example of all tests passing.
```
% DOMINO_API_HOST=<redacted> DOMINO_USER_API_KEY=<redacted> DOMINO_TOKEN_FILE='token_file.txt' pytest -sv tests/test_basic_auth.py
====================================== test session starts ======================================
cachedir: .pytest_cache
rootdir: /path/to/python-domino, configfile: pytest.ini
plugins: requests-mock-1.9.3, anyio-3.1.0
collected 4 items

tests/test_basic_auth.py::test_object_creation_with_api_key PASSED
tests/test_basic_auth.py::test_object_creation_with_token_file PASSED
tests/test_basic_auth.py::test_auth_against_real_deployment_with_api_key PASSED
tests/test_basic_auth.py::test_auth_against_real_deployment_with_token_file PASSED
```
And an example of a test run with skipped tests.
```
% pytest -sv tests/test_basic_auth.py
====================================== test session starts ======================================
cachedir: .pytest_cache
rootdir: /path/to/python-domino, configfile: pytest.ini
plugins: requests-mock-1.9.3, anyio-3.1.0
collected 4 items

tests/test_basic_auth.py::test_object_creation_with_api_key PASSED
tests/test_basic_auth.py::test_object_creation_with_token_file PASSED
tests/test_basic_auth.py::test_auth_against_real_deployment_with_api_key SKIPPED (No API key in environment)
tests/test_basic_auth.py::test_auth_against_real_deployment_with_token_file SKIPPED (No token file in environment)
```

dknupp commented:
@ddl-alexpanin -- I think this code is ready for you or someone on your team to merge.

dknupp commented:
@mikesbrown -- somehow I missed your last comments re: the domain name. Those are now addressed.
```
% whois somefakecompany.com | grep domain
domain:       COM
No match for domain "SOMEFAKECOMPANY.COM".
```
